sql/catalyst/src/main/scala/org/apache/spark/sql/util/ReuseMap.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.util
+
+import scala.collection.mutable.Map
+import scala.language.existentials
+
+import org.apache.spark.sql.catalyst.plans.QueryPlan
+import org.apache.spark.sql.types.StructType
+
+/**
+ * Map of canonicalized plans that can be used to find reuse possibilities.
+ *
+ * To avoid costly canonicalization of a plan:
+ * - we use its schema first to check if it can be replaced to a reused one at all
+ * - we insert it into the map of canonicalized plans only when at least 2 have the same schema
+ */
+class ReuseMap[T <: QueryPlan[_]] {
+  // scalastyle:off structural.type
+  private val map = Map[StructType, (T, Map[T2 forSome { type T2 >: T }, T])]()
+  // scalastyle:on structural.type
+
+  /**
+   * Find a matching plan with the same canonicalized form in the map or add the new plan to the
+   * map otherwise.
+   *
+   * @param plan the input plan
+   * @return the matching plan or the input plan
+   */
+  def lookupOrElseAdd(plan: T): T = {
+    val (firstSameSchemaPlan, sameResultPlans) = map.getOrElseUpdate(plan.schema, plan -> Map())
```

Comment:
The tuple of `plan -> Map()` here is to avoid inserting the plan into the map immediately to as to avoid calling`.canonicalized` if not necessary. Think of this as a lazy initialized map as once we have a plan that has the same schema we insert the first member of the tuple into the map.

Reply:
This is the 1. point in the description and tested by the case `SPARK-32041: No reuse interference inside ReuseExchange` in the new `ReuseExchangeAndSubquerySuite`: https://github.com/apache/spark/pull/28885/files#diff-f6f54d5cfc4254d8ed9122013394351bR28

Reply:
I see your point and I'm fine with reverting this change to the `ArrayBuffer`.

To me `Map[StructType, ArrayBuffer[SparkPlan]]` doesn't look intuitive at all exactly because of (a). Plan nodes are immutable case classes which provide consistent `equals()` and `hashcode()` and IMHO in those cases a map is better fit to look up an item. Using the `ArrayBuffer` suggests that something is weird with those nodes but I don't think it is. (We use canonicalized map of plans in `ReuseAdaptiveSubquery` too.)
To answer (b) as well, no, they very likely doesn't bring much improvement in a real word use case. The change is mainly due to the above reason.

Reply:
Commit https://github.com/apache/spark/pull/28885/commits/5df4c53d8f2805699bbcc2eae99380b59fd3ac3d reverts `ReuseMap` backing data structure to `Map[StructType, ArrayBuffer[Spark]]`.